nexus/src/app/external_endpoints.rs

```diff
@@ -488,69 +490,61 @@ pub(crate) async fn read_all_endpoints(
     datastore: &DataStore,
     opctx: &OpContext,
 ) -> Result<ExternalEndpoints, Error> {
-    // We will not look for more than this number of external DNS zones, Silos,
-    // or certificates.  We do not expect very many of any of these objects.
-    const MAX: u32 = 200;
-    let pagparams_id = DataPageParams {
-        marker: None,
-        limit: NonZeroU32::new(MAX).unwrap(),
-        direction: dropshot::PaginationOrder::Ascending,
-    };
-    let pagbyid = PaginatedBy::Id(pagparams_id);
-    let pagparams_name = DataPageParams {
-        marker: None,
-        limit: NonZeroU32::new(MAX).unwrap(),
-        direction: dropshot::PaginationOrder::Ascending,
-    };
-
-    let silos =
-        datastore.silos_list(opctx, &pagbyid, Discoverability::All).await?;
-    let external_dns_zones = datastore
-        .dns_zones_list(opctx, DnsGroup::External, &pagparams_name)
-        .await?;
+    // The batch size here is pretty arbitrary.  On the vast majority of
+    // systems, there will only ever be a handful of any of these objects.  Some
+    // systems are known to have a few dozen silos and a few hundred TLS
+    // certificates.  This code path is not particularly latency-sensitive.  Our
+    // purpose in limiting the batch size is just to avoid unbounded-size
+    // database transactions.
+    //
+    // unwrap(): safe because 200 is non-zero.
+    let batch_size = NonZeroU32::new(200).unwrap();
+
+    // Fetch all silos.
+    let mut silos = Vec::new();
+    let mut paginator = Paginator::new(batch_size);
+    while let Some(p) = paginator.next() {
+        let batch = datastore
+            .silos_list(
+                opctx,
+                &PaginatedBy::Id(p.current_pagparams()),
+                Discoverability::All,
+            )
+            .await?;
+        paginator = p.found_batch(&batch, &|s: &Silo| s.id());
+        silos.extend(batch.into_iter());
+    }
+
+    // Fetch all external DNS zones.  We should really only ever have one, but
+    // we may as well paginate this.
+    let mut external_dns_zones = Vec::new();
+    let mut paginator = Paginator::new(batch_size);
+    while let Some(p) = paginator.next() {
+        let batch = datastore
+            .dns_zones_list(opctx, DnsGroup::External, &p.current_pagparams())
+            .await?;
+        paginator = p.found_batch(&batch, &|z: &DnsZone| z.zone_name.clone());
```

Comment:
This all looks correct to me, but out of curiosity: why in the other two queries do we have an explicit `PaginatedBy::Id(pageparams)`, and for this one we seem to be implicitly paginating by zone name? Is the use of `z.zone_name` here relying on implementation details of `dns_zones_list()`?

Reply:
If I recall correctly: the sort of low-level parameter for pagination is `DataPageParams`, which can be parametrized by basically any marker type:
https://github.com/oxidecomputer/omicron/blob/460f038ccd57fead6a78b01ae9bf8b4fee38d54f/common/src/api/external/mod.rs#L95-L120

For collections that only support pagination by one thing, like `dns_zones_list()`, their datastore functions just accept `DataPageParams`:
https://github.com/oxidecomputer/omicron/blob/460f038ccd57fead6a78b01ae9bf8b4fee38d54f/nexus/db-queries/src/db/datastore/dns.rs#L52

For relatively few collections we support paginating by any one of a few different fields (well, it's always "id" _or_ "name").  The caller gets to pick.  The code is basically the same and these functions accept a `PaginatedBy`, which is just an enum with variants for by-id or by-name, and each variant just contains the appropriate `DataPageParams` type:
https://github.com/oxidecomputer/omicron/blob/460f038ccd57fead6a78b01ae9bf8b4fee38d54f/common/src/api/external/http_pagination.rs#L364-L368

So I don't think this is relying on an implementation detail.  It's just that `dns_zones_list()` only supports paginating by the zone name so it accepts `DataPageParams` directly, while the other functions support pagination by either and so need the wrapper that lets you pick which one you want (but ultimately you're still providing a `DataPageParams` and that's what the code is using).

--- 

There's a bit of an implicit detail here which is that the caller of a paginated function needs to know which field of the object is the marker.  (In practice this is not super easy to get wrong because the marker _type_ is usually kind of unique -- a uuid or a Name -- though you could pick a different field that's also a uuid or something.)  This applies to all three queries though and every other use of `Paginator`, too (it's encoded in the closure provided to `found_batch()`).